examples/ecs/observers.rs

```diff
@@ -0,0 +1,199 @@
+                (mine.pos.x / CELL_SIZE).floor() as i32,
+                (mine.pos.y / CELL_SIZE).floor() as i32,
+            );
+            index.map.entry(tile).or_default().insert(observer.source());
+        },
+    );
+
+    // Since observers run systems you can also define them as standalone functions rather than closures.
+    // This observer runs whenever the `Mine` component is removed from an entity (including despawning it)
+    // and removes it from the spatial index.
+    world.observer(remove_mine);
+
+    // Now we spawn a set of random mines.
+    for _ in 0..1000 {
+        world
+            .spawn(Mine {
+                pos: Vec2::new(
+                    (rand::random::<f32>() - 0.5) * 1200.0,
+                    (rand::random::<f32>() - 0.5) * 600.0,
+                ),
+                size: 4.0 + rand::random::<f32>() * 16.0,
+            })
+            // Observers can also listen to events targeting a specific entity.
+            // This observer listens to `Explode` events targeted at our mine.
+            .observe(
+                |observer: Observer<Explode>, query: Query<&Mine>, mut commands: Commands| {
+                    // If an event is targeting a specific entity you can access it with `.source()`
+                    let source = observer.source();
+                    let Some(mut entity) = commands.get_entity(source) else {
+                        return;
+                    };
+                    println!("Boom! {:?} exploded.", source.index());
+                    entity.despawn();
+                    let mine = query.get(source).unwrap();
+                    // Fire another event to cascade into other mines.
+                    commands
+                        .event(TriggerMines {
+                            pos: mine.pos,
+                            radius: mine.size,
+                        })
+                        .emit();
+                },
+            );
+    }
+}
+
+#[derive(Resource, Default)]
+struct SpatialIndex {
+    map: HashMap<(i32, i32), HashSet<Entity>>,
+}
+
+/// Cell size has to be bigger than any `TriggerMine::radius`
+const CELL_SIZE: f32 = 64.0;
+
+impl SpatialIndex {
+    // Lookup all entities within adjacent cells of our spatial index
+    fn get_nearby(&self, pos: Vec2) -> Vec<Entity> {
+        let tile = (
+            (pos.x / CELL_SIZE).floor() as i32,
+            (pos.y / CELL_SIZE).floor() as i32,
+        );
+        let mut nearby = Vec::new();
+        for x in -1..2 {
+            for y in -1..2 {
+                if let Some(mines) = self.map.get(&(tile.0 + x, tile.1 + y)) {
+                    nearby.extend(mines.iter());
+                }
+            }
+        }
+        nearby
+    }
+}
+
+// Remove despawned mines from our index
+fn remove_mine(
+    observer: Observer<OnRemove, Mine>,
+    query: Query<&Mine>,
+    mut index: ResMut<SpatialIndex>,
+) {
+    let mine = query.get(observer.source()).unwrap();
```

Comment:
Is it safe to unwrap here? If so, a comment explaining why it's safe would be good.

Reply:
Yes in the sense that the entity should always have the component during execution of `OnRemove` however technically this requires the command ordering PR to be bulletproof in the case of other observers listening to the same trigger.